File: src/go/phonemize/japanese.go
package phonemize

import (
	"regexp"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// JapaneseG2PEngine — interface for fullcontext label extraction
// ---------------------------------------------------------------------------

// JapaneseG2PEngine provides fullcontext labels from text.
// Phase 1 uses a subprocess backend (e.g. open_jtalk CLI).
type JapaneseG2PEngine interface {
	ExtractFullcontext(text string) ([]string, error)
}

// ---------------------------------------------------------------------------
// JapanesePhonemizer
// ---------------------------------------------------------------------------

// JapanesePhonemizer converts Japanese text to phonemes using the Kurihara
// method. It delegates fullcontext label extraction to a JapaneseG2PEngine.
type JapanesePhonemizer struct {
	engine JapaneseG2PEngine
}

// NewJapanesePhonemizer creates a JapanesePhonemizer backed by the given engine.
func NewJapanesePhonemizer(engine JapaneseG2PEngine) *JapanesePhonemizer {
	return &JapanesePhonemizer{engine: engine}
}

// PhonemizeWithProsody converts Japanese text to phoneme tokens with prosody.
// It implements the Phonemizer interface.
func (p *JapanesePhonemizer) PhonemizeWithProsody(text string) (*PhonemizeResult, error) {
	labels, err := p.engine.ExtractFullcontext(text)
	if err != nil {
		return nil, err
	}
	return labelsToTokensWithProsody(labels, text)
}

// LanguageCode returns "ja".
func (p *JapanesePhonemizer) LanguageCode() string {
	return "ja"
}

// ---------------------------------------------------------------------------
// Label parsing regexes
// ---------------------------------------------------------------------------

var (
	rePhoneme = regexp.MustCompile(`-([^+]+)\+`)
	reA1      = regexp.MustCompile(`/A:([0-9\-]+)\+`)
	reA2      = regexp.MustCompile(`\+(\d+)\+`)
	reA3      = regexp.MustCompile(`\+(\d+)/`)
)

// ---------------------------------------------------------------------------
// Core: fullcontext labels -> phoneme tokens with prosody (Kurihara method)
// ---------------------------------------------------------------------------

// labelsToTokensWithProsody converts OpenJTalk fullcontext labels to phoneme
// tokens with prosody information using the Kurihara method.
//
// The Kurihara method inserts the following prosody marks:
//   - ^   : beginning of sentence (BOS)
//   - $/? : end of sentence (EOS / question type)
//   - _   : short pause (pau)
//   - #   : accent phrase boundary
//   - [   : rising pitch mark (accent phrase head)
//   - ]   : falling pitch mark (accent nucleus)
func labelsToTokensWithProsody(labels []string, text string) (*PhonemizeResult, error) {
	var tokens []string
	var prosody []*ProsodyInfo

	numLabels := len(labels)

	for idx, label := range labels {
		m := rePhoneme.FindStringSubmatch(label)
		if m == nil {
			continue
		}
		phoneme := m[1]

		// Beginning / end silence
		if phoneme == "sil" {
			switch idx {
			case 0:
				tokens = append(tokens, "^")
				prosody = append(prosody, nil)
			case numLabels - 1:
				qt := getQuestionType(text)
				tokens = append(tokens, qt)
				prosody = append(prosody, nil)
			}
			continue
		}

		// Short pause
		if phoneme == "pau" {
			tokens = append(tokens, "_")
			prosody = append(prosody, nil)
			continue
		}

		// Regular phoneme
		tokens = append(tokens, phoneme)

		// Extract A1/A2/A3
		mA1 := reA1.FindStringSubmatch(label)
		mA2 := reA2.FindStringSubmatch(label)
		mA3 := reA3.FindStringSubmatch(label)

		if mA1 == nil || mA2 == nil || mA3 == nil {
			prosody = append(prosody, nil)
			continue
		}

		a1, _ := strconv.Atoi(mA1[1])
		a2, _ := strconv.Atoi(mA2[1])
		a3, _ := strconv.Atoi(mA3[1])

		prosody = append(prosody, &ProsodyInfo{A1: a1, A2: a2, A3: a3})

		// Look-ahead: fetch a2 of the next label
		a2Next := -1
		if idx < numLabels-1 {
			mA2Next := reA2.FindStringSubmatch(labels[idx+1])
			if mA2Next != nil {
				a2Next, _ = strconv.Atoi(mA2Next[1])
			}
		}

		// Insert accent nucleus mark "]" (pitch H->L)
		if a1 == 0 && a2Next == a2+1 {
			tokens = append(tokens, "]")
			prosody = append(prosody, nil)
		}

		// Insert accent phrase boundary "#"
		if a2 == a3 && a2Next == 1 {
			tokens = append(tokens, "#")
			prosody = append(prosody, nil)
		}

		// Insert rising pitch mark "["
		if a2 == 1 && a2Next == 2 {
			tokens = append(tokens, "[")
			prosody = append(prosody, nil)
		}
	}

	// Apply context-dependent N phoneme rules
	tokens = applyNPhonemeRules(tokens)

	// Map multi-char tokens to PUA codepoints
	tokens = MapSequence(tokens)

	// Determine EOS token
	eosToken := getQuestionType(text)

	return &PhonemizeResult{
		Tokens:   tokens,
		Prosody:  prosody,
		EOSToken: eosToken,
	}, nil
}

// ---------------------------------------------------------------------------
// Question type detection
// ---------------------------------------------------------------------------

// getQuestionType returns the appropriate end-of-sentence marker based on
// the text ending. Returns one of: "?!", "?.", "?~", "?", or "$".
func getQuestionType(text string) string {
	stripped := strings.TrimSpace(text)

	// Multi-char patterns first (longer before shorter)
	if strings.HasSuffix(stripped, "?!") ||
		strings.HasSuffix(stripped, "\uff01\uff1f") || // ！？
		strings.HasSuffix(stripped, "\uff1f\uff01") { // ？！
		return "?!"
	}
	if strings.HasSuffix(stripped, "?.") ||
		strings.HasSuffix(stripped, "\u3002\uff1f") || // 。？
		strings.HasSuffix(stripped, "\uff1f\u3002") { // ？。
		return "?."
	}
	if strings.HasSuffix(stripped, "?~") ||
		strings.HasSuffix(stripped, "\uff5e\uff1f") || // ～？
		strings.HasSuffix(stripped, "\uff1f\uff5e") { // ？～
		return "?~"
	}

	// Single ? fallback
	if strings.HasSuffix(stripped, "?") || strings.HasSuffix(stripped, "\uff1f") {
		return "?"
	}

	return "$"
}

// ---------------------------------------------------------------------------
// Context-dependent N phoneme rules
// ---------------------------------------------------------------------------

// skipTokens is the set of tokens that should be skipped when looking ahead
// for the next phoneme after "N".
var skipTokens = map[string]bool{
	"_":  true,
	"#":  true,
	"[":  true,
	"]":  true,
	"^":  true,
	"$":  true,
	"?":  true,
	"?!": true,
	"?.": true,
	"?~": true,
}

// bilabial phonemes: m, my, b, by, p, py
var bilabials = map[string]bool{
	"m": true, "my": true,
	"b": true, "by": true,
	"p": true, "py": true,
}

// alveolar phonemes: n, ny, t, ty, d, dy, ts, ch
var alveolars = map[string]bool{
	"n": true, "ny": true,
	"t": true, "ty": true,
	"d": true, "dy": true,
	"ts": true, "ch": true,
}

// velar phonemes: k, ky, kw, g, gy, gw
var velars = map[string]bool{
	"k": true, "ky": true, "kw": true,
	"g": true, "gy": true, "gw": true,
}

// applyNPhonemeRules replaces each "N" token with a context-dependent variant:
//   - N_m      : before bilabial (m, my, b, by, p, py)
//   - N_n      : before alveolar (n, ny, t, ty, d, dy, ts, ch)
//   - N_ng     : before velar (k, ky, kw, g, gy, gw)
//   - N_uvular : at phrase end or before vowels/other consonants
//
// This is a 1-to-1 replacement (each "N" maps to exactly one variant token),
// so the token count is unchanged and the prosody array stays aligned.
func applyNPhonemeRules(tokens []string) []string {
	result := make([]string, 0, len(tokens))

	for i, token := range tokens {
		if token != "N" {
			result = append(result, token)
			continue
		}

		// Look ahead past skip tokens to find next phoneme
		var nextPhoneme string
		for j := i + 1; j < len(tokens); j++ {
			if !skipTokens[tokens[j]] {
				nextPhoneme = tokens[j]
				break
			}
		}

		switch {
		case nextPhoneme == "":
			result = append(result, "N_uvular") // end of phrase
		case bilabials[nextPhoneme]:
			result = append(result, "N_m")
		case alveolars[nextPhoneme]:
			result = append(result, "N_n")
		case velars[nextPhoneme]:
			result = append(result, "N_ng")
		default:
			result = append(result, "N_uvular") // vowels, other consonants
		}
	}

	return result
}

// ---------------------------------------------------------------------------
// Compile-time interface check
// ---------------------------------------------------------------------------

// Verify that *JapanesePhonemizer satisfies the Phonemizer interface.
var _ Phonemizer = (*JapanesePhonemizer)(nil)

File: src/python/piper_train/phonemize/japanese.py
import re


# Try to import pyopenjtalk-plus first (Windows compatible), fall back to pyopenjtalk
try:
    import pyopenjtalk_plus as pyopenjtalk
except ImportError:
    try:
        import pyopenjtalk
    except ImportError:
        raise ImportError(
            "Neither pyopenjtalk nor pyopenjtalk-plus is installed"
        ) from None

from .base import Phonemizer, ProsodyInfo
from .custom_dict import CustomDictionary
from .token_mapper import map_sequence


__all__ = [
    "phonemize_japanese",
    "phonemize_japanese_with_prosody",
    "ProsodyInfo",
    "JapanesePhonemizer",
]


# Regular expressions reused many times
_RE_PHONEME = re.compile(r"-([^+]+)\+")
_RE_A1 = re.compile(r"/A:([\d-]+)\+")
_RE_A2 = re.compile(r"\+([0-9]+)\+")
_RE_A3 = re.compile(r"\+([0-9]+)/")


def _is_question(text: str) -> bool:
    """Return True if *text* ends with a Japanese/ASCII question mark."""
    return text.strip().endswith("?") or text.strip().endswith("？")


def _get_question_type(text: str) -> str:
    """Return the appropriate question marker based on text ending.

    Returns one of: "?!", "?.", "?~", "?", or "$" (for non-questions).

    Markers:
    - "?!" : Emphatic question (強調疑問) - ends with ?! or ！？
    - "?." : Neutral/rhetorical question (平叙疑問) - ends with ?. or 。？
    - "?~" : Tag question (確認疑問) - ends with ?~ or ～？ or ？～
    - "?"  : Generic question - ends with ? or ？
    - "$"  : Declarative (non-question)
    """
    stripped = text.strip()

    # Multi-char patterns first (check longer patterns before shorter)
    if (
        stripped.endswith("?!")
        or stripped.endswith("！？")
        or stripped.endswith("？！")
    ):
        return "?!"
    if (
        stripped.endswith("?.")
        or stripped.endswith("。？")
        or stripped.endswith("？。")
    ):
        return "?."
    if (
        stripped.endswith("?~")
        or stripped.endswith("～？")
        or stripped.endswith("？～")
    ):
        return "?~"

    # Single ? fallback
    if stripped.endswith("?") or stripped.endswith("？"):
        return "?"

    return "$"  # Not a question


# Set of tokens that should be skipped when looking for next phoneme
_SKIP_TOKENS = frozenset(("_", "#", "[", "]", "^", "$", "?", "?!", "?.", "?~"))


def _apply_n_phoneme_rules(tokens: list[str]) -> list[str]:
    """Apply context-dependent rules to replace 'N' with specific variants.

    Japanese 'ん' (N) has different pronunciations depending on the following
    phoneme:
    - N_m     : before m/b/p (bilabial assimilation)
    - N_n     : before n/t/d/ts/ch (alveolar assimilation)
    - N_ng    : before k/g (velar assimilation)
    - N_uvular: at phrase end or before vowels/other consonants

    Parameters
    ----------
    tokens : list[str]
        List of phoneme tokens (before map_sequence)

    Returns
    -------
    list[str]
        List with 'N' replaced by context-appropriate variants
    """
    result = []
    for i, token in enumerate(tokens):
        if token != "N":
            result.append(token)
            continue

        # Look ahead to find next actual phoneme
        next_phoneme = None
        for j in range(i + 1, len(tokens)):
            if tokens[j] not in _SKIP_TOKENS:
                next_phoneme = tokens[j]
                break

        # Determine N variant based on next phoneme
        if next_phoneme is None:
            result.append("N_uvular")  # End of phrase
        elif next_phoneme in ("m", "my", "b", "by", "p", "py"):
            result.append("N_m")  # Bilabial
        elif next_phoneme in ("n", "ny", "t", "ty", "d", "dy", "ts", "ch"):
            result.append("N_n")  # Alveolar
        elif next_phoneme in ("k", "ky", "kw", "g", "gy", "gw"):
            result.append("N_ng")  # Velar
        else:
            result.append("N_uvular")  # Vowels, other consonants

    return result


def phonemize_japanese(
    text: str, custom_dict: CustomDictionary | str | list[str] | None = None
) -> list[str]:
    """Convert *text* into a list of phoneme/prosody tokens that Piper can ingest.

    The algorithm follows the so-called "Kurihara method" that inserts the
    following extra symbols in the phoneme sequence:

    ^   : beginning of sentence
    $/?: end of sentence (choose ? for interrogative)
    _   : short pause (pau)
    #   : accent phrase boundary
    [   : rising-pitch mark (accent phrase head)
    ]   : falling-pitch mark (accent nucleus)

    Parameters
    ----------
    text : str
        Input text to phonemize
    custom_dict : CustomDictionary, str, List[str], optional
        Custom dictionary instance or path(s) to dictionary file(s)

    Notes
    -----
    1. We rely on *pyopenjtalk.extract_fullcontext* to obtain full-context labels.
    2. "sil" at the beginning / end of the utterance is converted into ^ / $ or ?.
    3. Custom dictionary is applied before OpenJTalk processing for better pronunciation.
    """

    # カスタム辞書を適用
    if custom_dict is not None:
        if isinstance(custom_dict, CustomDictionary):
            dictionary = custom_dict
        else:
            # パスが渡された場合は辞書を作成
            dictionary = CustomDictionary(custom_dict)

        # テキストに辞書を適用
        text = dictionary.apply_to_text(text)

    labels = pyopenjtalk.extract_fullcontext(text)
    tokens: list[str] = []

    for idx, label in enumerate(labels):
        m_ph = _RE_PHONEME.search(label)
        if not m_ph:
            # Should never happen – skip just in case
            continue
        phoneme = m_ph.group(1)

        # Beginning / end silence handling
        if phoneme == "sil":
            if idx == 0:
                tokens.append("^")
            elif idx == len(labels) - 1:
                # Always add end marker when we find the last sil
                tokens.append(_get_question_type(text))
            # Skip adding ordinary phoneme for sil
            continue

        # Short pause
        if phoneme == "pau":
            tokens.append("_")
            continue

        # Keep unvoiced vowels as uppercase for linguistic accuracy

        tokens.append(phoneme)

        # ------------------------------------------------------------------
        # Prosody mark extraction – see Open JTalk label definition
        # ------------------------------------------------------------------
        # A1 : accented? 1 if accented mora else 0
        # A2 : position of current mora in the accent phrase (1-based)
        # A3 : number of mora in the accent phrase
        #
        # A2_next is needed to detect accent nucleus and phrase boundary.
        # ------------------------------------------------------------------
        m_a1 = _RE_A1.search(label)
        m_a2 = _RE_A2.search(label)
        m_a3 = _RE_A3.search(label)
        if not (m_a1 and m_a2 and m_a3):
            # Cannot get accent info – continue
            continue

        a1 = int(m_a1.group(1))
        a2 = int(m_a2.group(1))
        a3 = int(m_a3.group(1))

        # Look-ahead to next label to fetch a2_next
        if idx < len(labels) - 1:
            m_a2_next = _RE_A2.search(labels[idx + 1])
            a2_next = int(m_a2_next.group(1)) if m_a2_next else -1
        else:
            a2_next = -1

        # Insert accent nucleus mark "]" at the descending point.
        # Kurihara rule: a1==0 && a2_next == a2 + 1 (i.e., pitch goes from H to L)
        if (a1 == 0) and (a2_next == a2 + 1):
            tokens.append("]")

        # Insert accent phrase boundary "#" when current mora is last in phrase
        if (a2 == a3) and (a2_next == 1):
            tokens.append("#")

        # Insert rising mark "[" at phrase head (a2==1) when next mora is 2
        if (a2 == 1) and (a2_next == 2):
            tokens.append("[")

    # Apply context-dependent N phoneme rules
    tokens = _apply_n_phoneme_rules(tokens)

    # 多文字トークンを1コードポイントへ変換
    return map_sequence(tokens)


def phonemize_japanese_with_prosody(
    text: str, custom_dict: CustomDictionary | str | list[str] | None = None
) -> tuple[list[str], list[ProsodyInfo | None]]:
    """Convert *text* into phoneme tokens with prosody information.

    This function extends phonemize_japanese() by also returning A1/A2/A3
    prosody values from OpenJTalk labels for each phoneme token.

    Parameters
    ----------
    text : str
        Input text to phonemize
    custom_dict : CustomDictionary, str, List[str], optional
        Custom dictionary instance or path(s) to dictionary file(s)

    Returns
    -------
    tuple[list[str], list[ProsodyInfo | None]]
        A tuple containing:
        - tokens: List of phoneme/prosody tokens (same as phonemize_japanese)
        - prosody_info: List of ProsodyInfo for each token, or None for
          special tokens (^, $, ?, _, #, [, ])

    Notes
    -----
    The prosody information (A1/A2/A3) is useful for:
    - A1: Detecting accent nucleus position
    - A2: Position-aware duration prediction
    - A3: Phrase-level prosody control

    Example
    -------
    >>> tokens, prosody = phonemize_japanese_with_prosody("こんにちは")
    >>> for t, p in zip(tokens, prosody):
    ...     if p:
    ...         print(f"{t}: A1={p.a1}, A2={p.a2}, A3={p.a3}")
    """
    # カスタム辞書を適用
    if custom_dict is not None:
        if isinstance(custom_dict, CustomDictionary):
            dictionary = custom_dict
        else:
            dictionary = CustomDictionary(custom_dict)
        text = dictionary.apply_to_text(text)

    labels = pyopenjtalk.extract_fullcontext(text)
    tokens: list[str] = []
    prosody_info: list[ProsodyInfo | None] = []

    for idx, label in enumerate(labels):
        m_ph = _RE_PHONEME.search(label)
        if not m_ph:
            continue
        phoneme = m_ph.group(1)

        # Beginning / end silence handling
        if phoneme == "sil":
            if idx == 0:
                tokens.append("^")
                prosody_info.append(None)
            elif idx == len(labels) - 1:
                tokens.append(_get_question_type(text))
                prosody_info.append(None)
            continue

        # Short pause
        if phoneme == "pau":
            tokens.append("_")
            prosody_info.append(None)
            continue

        # Add phoneme token
        tokens.append(phoneme)

        # Extract A1/A2/A3 values
        m_a1 = _RE_A1.search(label)
        m_a2 = _RE_A2.search(label)
        m_a3 = _RE_A3.search(label)

        if m_a1 and m_a2 and m_a3:
            a1 = int(m_a1.group(1))
            a2 = int(m_a2.group(1))
            a3 = int(m_a3.group(1))
            prosody_info.append(ProsodyInfo(a1=a1, a2=a2, a3=a3))

            # Look-ahead for prosody marks
            if idx < len(labels) - 1:
                m_a2_next = _RE_A2.search(labels[idx + 1])
                a2_next = int(m_a2_next.group(1)) if m_a2_next else -1
            else:
                a2_next = -1

            # Insert accent nucleus mark "]"
            if (a1 == 0) and (a2_next == a2 + 1):
                tokens.append("]")
                prosody_info.append(None)

            # Insert accent phrase boundary "#"
            if (a2 == a3) and (a2_next == 1):
                tokens.append("#")
                prosody_info.append(None)

            # Insert rising mark "["
            if (a2 == 1) and (a2_next == 2):
                tokens.append("[")
                prosody_info.append(None)
        else:
            # No prosody info available
            prosody_info.append(None)

    # Apply context-dependent N phoneme rules
    # Note: This only replaces 'N' with variants in-place, so prosody_info alignment is preserved
    tokens = _apply_n_phoneme_rules(tokens)

    # Map multi-character tokens to single codepoints
    mapped_tokens = map_sequence(tokens)

    return mapped_tokens, prosody_info


class JapanesePhonemizer(Phonemizer):
    """Japanese phonemizer using OpenJTalk."""

    def phonemize(self, text: str) -> list[str]:
        return phonemize_japanese(text)

    def phonemize_with_prosody(
        self, text: str
    ) -> tuple[list[str], list[ProsodyInfo | None]]:
        return phonemize_japanese_with_prosody(text)

    def get_phoneme_id_map(self) -> dict[str, list[int]] | None:
        return None

    def post_process_ids(
        self,
        phoneme_ids: list[int],
        prosody_features: list[dict | None],
        phoneme_id_map: dict[str, list[int]],
    ) -> tuple[list[int], list[dict | None]]:
        """No-op: Japanese handles BOS/EOS/padding inline during phonemization.

        For multilingual/bilingual models, the caller should use
        MultilingualPhonemizer which adds intersperse padding in its
        own post_process_ids.
        """
        return phoneme_ids, prosody_features
